src/scene.rs

```diff
@@ -28,4 +29,115 @@ pub struct Scene {
+/// # fn main() {}
+/// ```
+///
+/// You need this wrapper around `three::Mesh` to cache some information - in our case, visibility.
+///
+/// In your game you contain all your enemy objects in `Vec<Enemy>`. In the main loop you need
+/// to iterate over all the enemies and make them visible or not, basing on current position.
+/// The most obvious way is to use [`Object::sync`], but it's not the best idea from the side of
+/// performance. Istead, you can create `SyncGuard` and use its `update` method to effectively
+/// walk through every enemy in your game:
+///
+/// ```rust,no_run
+/// # #[macro_use]
+/// # extern crate three;
+/// # #[derive(Clone)]
+/// # struct Enemy {
+/// #     mesh: three::Mesh,
+/// #     is_visible: bool,
+/// # }
+/// # three_object_wrapper!(Enemy::mesh);
+/// # fn main() {
+/// # let mut win = three::Window::new("SyncGuard example");
+/// # let geometry = three::Geometry::default();
+/// # let material = three::material::Basic { color: three::color::RED, map: None };
+/// # let mesh = win.factory.mesh(geometry, material);
+/// # let mut enemy = Enemy { mesh, is_visible: true };
+/// # enemy.set_parent(&win.scene);
+/// # let mut enemies = vec![enemy];
+/// # while true {
+/// let mut sync = win.scene.with_sync();
+/// sync.update(enemies.iter_mut(), |enemy, node| {
+///     let position = node.world_transform.position;
+///     if position.x > 10.0 {
+///         enemy.is_visible = false;
+///         enemy.set_visible(false);
+///     } else {
+///         enemy.is_visible = true;
+///         enemy.set_visible(true);
+///     }
+/// });
+/// # }}
+/// ```
+///
+/// [`Object::sync`]: ../struct.Object.html#method.sync
+pub struct SyncGuard<'a> {
+    pub(crate) scene: &'a mut Scene,
+    pub(crate) scene_id: Option<Uid>,
+}
+
+impl<'a> SyncGuard<'a> {
+    /// Walk through every item in `iterator`, obtain its [`Node`] and call user-supplied closure.
+    ///
+    /// # Panics
+    /// Panics if `scene` doesn't have this `Object`.
+    ///
+    /// [`Node`]: ../node/struct.Node.html
+    pub fn update<T: 'a, I, F>(
+        &mut self,
+        iterator: I,
+        mut closure: F,
+    ) where
+        T: AsRef<Object>,
+        I: Iterator<Item = &'a mut T>,
+        F: FnMut(&'a mut T, Node),
+    {
+        let mut hub = self.scene.hub.lock().unwrap();
+        iterator.for_each(|item| {
+            let node = {
+                let object: &Object = item.as_ref();
+                let node = &hub.nodes[&object.node];
+                assert_eq!(node.scene_id, self.scene_id);
+                node.into()
+            };
+            closure(item, node);
+        })
+    }
+}
+
+impl Scene {
+    /// Create new [`SyncGuard`](struct.SyncGuard.html).
+    ///
+    /// This is performance-costly operation, you should not use it many times per frame.
+    pub fn with_sync<'a>(&'a mut self) -> SyncGuard<'a> {
+        let scene_id = {
+            let mut hub = self.hub.lock().unwrap();
```

Comment:
Doesn't matter too much at the moment but it might be worth keeping hold of the lock for the lifetime of `SyncGuard`.